### PR TITLE
Test skeleton

### PR DESCRIPTION
**Pull request by shammamah**
## About 
- [ ] This is a new component 
- [ ] I am adding a feature to an existing component, or improving an existing feature
- [ ] I am closing an issue 

## Description of changes 
Adding a basic test suite to ensure that all of our tests have similar formatting (and also to explain the interface for the common tests written by @Bachibouzouk). Piggybacks off #142 to make use of the common test suite. 

## Before merging
- [x] I have gone through the [code review checklist](https://github.com/plotly/dash-component-boilerplate/blob/master/%7B%7Bcookiecutter.project_shortname%7D%7D/review_checklist.md)
- [x] I have completed all of the [steps to take before merging](before_merging.md)
- [x] I know how to [deploy to DDS](deploying_to_dds.md)

### CODE REVIEW

**Comment by shammamah** [comment]
@mkcor We should ignore the skeleton file so that pylint doesn't fail all the time! Do you know how to do that?

File: tests/test_skeleton.py [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+             prop_name, after casting it to the correct type. 
+    """
+
+    # avoid triggering this callback when the button is first created
+    if nclicks is not None:
+        # cast the string representation of the desired prop value
+        # into the appropriate type
+
+        if prop_type in PROP_TYPES:
+            return PROP_TYPES[prop_type](prop_value)
```
Comment (Bachibouzouk) [inline tests/test_skeleton.py]: Using multiple `return` can make functions difficult to debug as mentioned [here](https://docs.python-guide.org/writing/style/#returning-values). I would be in favor to have a single `return` statement unless unavoidable.

File: tests/test_skeleton.py [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    template_test_component_single_prop(
+        dash_threaded,
+        selenium,
+        APP_NAME,
+        assert_callback,
+        COMPONENTNAME_test_props_callback,
+        PROPNAME,
+        input_PROPNAME,
+        propType=propType,
```
Comment (Bachibouzouk) [inline tests/test_skeleton.py]: it should be `prop_type` in snake case ;)

File: tests/test_skeleton.py (hunk 1 above)
Comment (shammamah) [reply tests/test_skeleton.py]: Fixed in cbc8bd0!

File: tests/test_skeleton.py (hunk 2 above)
Comment (shammamah) [reply tests/test_skeleton.py]: Fixed in 709f8eb!

**Comment by mkcor** [comment]
@shammamah in the CI config file, you want to modify [line](https://github.com/plotly/dash-bio/blob/bb76123831a55cdb462c2bdb2587ea8ae7f2ecf3/.circleci/config.yml#L77)
```bash
pytest --driver chrome tests
```
so it reads
```bash
pytest --driver chrome tests --ignore tests/test_skeleton.py
```
instead.

**Comment by mkcor** [comment]
@shammamah and then apply the same change to the corresponding line of `CONTRIBUTING.md` (about running tests locally).

File: tests/test_skeleton.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+
+# for React components, we need to define a way to interact with the
+# props directly (instead of through a graph component); to that end,
+# we define a callback method that defines how exactly a prop is
+# updated
+
+# this callback will be used in the simple test app, which consists of
+# the component, a single button, and two inputs
+
+def COMPONENTNAME_test_props_callback(
```
Comment (mkcor) [inline tests/test_skeleton.py]:
```suggestion
def test_COMPONENTNAME_props_callback(
```
I would prefer sticking with the convention that names of testing functions start with `test_` (we would have to update the function name of the volcano plot test).

File: tests/test_skeleton.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        'dash-bio/{}'.format(APP_NAME)
+
+    
+# below, write tests for initial conditions; they will most likely
+# make use of wait_for_text_to_equal and other similar functions
+
+@init_demo_app(APP_NAME)
+def test_INITIALCONDITION(dash_threaded, selenium):
+    """Test some initial condition here (e.g., that the correct dataset
+    has been loaded)."""
```
[the remark this answers is not in the document]
Comment (shammamah) [reply tests/test_skeleton.py]: Fixed in https://github.com/plotly/dash-bio/pull/150/commits/773b8c417a8e48a8cdbaf4d500c8a45baaf03491

File: tests/test_skeleton.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+
+# below, write tests for interactions with the app (e.g., selecting a
+# value from a dropdown or filling in some sort of input); they will
+# most likely make use of send_keys
+
+@init_demo_app(APP_NAME)
+def test_CHANGESOMETHING(dash_threaded, selenium):
+    """Test the results of changing something here (e.g., the value
+    of a dropdown)."""
```
[the remark this answers is not in the document]
Comment (shammamah) [reply tests/test_skeleton.py]: Fixed in https://github.com/plotly/dash-bio/pull/150/commits/773b8c417a8e48a8cdbaf4d500c8a45baaf03491

**Comment by shammamah** [comment]
@mkcor Still having some trouble with getting pylint/flake8/CI to ignore the skeleton file... any ideas?

**Comment by mkcor** [comment]
I'm curious, why do you have a merge commit on this branch which is not a merge of master (I mean d930874)?

File: tests/test_skeleton.py [hunk 6]
```diff
@@ -0,0 +1,159 @@
+# pylint: skip-file
```
[the remark this answers is not in the document]
Comment (shammamah) [reply tests/test_skeleton.py]: Fixed in 4919913

**Comment by shammamah** [comment]
It's because you can commit suggested changes directly from the web interface -- I did that for some of the suggestions you made. So, the remote branch was actually ahead of my local branch and I had to merge it back in!

File: .circleci/config.yml [hunk 7]
```diff
@@ -59,22 +59,22 @@ jobs:
                 name: Run pylint
                 command: |
                     . venv/bin/activate
-                    pylint tests/ dash_bio/utils/ dash_bio/component_factory/
+                    pylint tests/ dash_bio/utils/ dash_bio/component_factory/ --ignore tests/test_skeleton.py
```
Comment (mkcor) [inline .circleci/config.yml]: You don't want to make this change.

File: tests/test_skeleton.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
+import os
```
Comment (mkcor) [inline tests/test_skeleton.py]: Have you tried starting this file with line
```py
# pylint: disable=W,C,R
```
?

File: .circleci/config.yml [hunk 9]
```diff
@@ -59,22 +59,22 @@ jobs:
...
                 when: always
 
             - run:
                 name: Run flake8
                 command: |
                     . venv/bin/activate
-                    flake8 --max-line-length=100 tests/ dash_bio/utils/ dash_bio/component_factory/
+                    flake8 --max-line-length=100 tests/ dash_bio/utils/ dash_bio/component_factory/ --ignore tests/test_skeleton.py
```
Comment (mkcor) [inline .circleci/config.yml]: You don't want to make this change either: With `flake8`, the `--ignore` option is for specifying error codes, not files.
Reference: http://flake8.pycqa.org/en/latest/user/options.html#cmdoption-flake8-ignore

File: tests/test_skeleton.py (hunk 8 above)
Comment (Bachibouzouk) [reply tests/test_skeleton.py]: @shammamah, @mkcor : I tried and it worked (I use pylint on it locally and then added the line `# pylint: disable=E,W,C,R`)

**Pull request merged**